_posts/2021-04-19-solving-problems-with-custom-extensions.adoc

```diff
@@ -0,0 +1,168 @@
+
+I could write an article to debunk these arguments and explain how the latter makes your development workflow much more efficient and how the former makes the latter possible, even if fast boots are not your thing.
+But for the sake of this blog post, let's admit these persons are absolutely right and they are not good reasons to use Quarkus.
+
+So now what? Back to `<insert your favorite framework here>`? Not so fast.
+
+Quarkus didn't achieve fast startup and low memory footprint by using dark magic or lazy loading tricks but by rethinking entirely the way Java applications are bootstrapped.
+The whole point of Quarkus is to move as much work as possible to the build and this journey made us build a framework to push work at build time and this framework can be leveraged in Quarkus extensions.
+
+> A Quarkus extension? That sounds like a lot of work?
+
+No, really, it is not.
+You can develop your own extensions very easily and they can solve some unordinary issues in a very simple way.
+
+Last week, one of of our users (hey, Juan!) asked this question on Zulip:
+
+> Hi! I'm trying to understand how to find classes with some criteria and add them to the dependency injection context, for example: I want to find all classes whose name ends with "MessageTransformer" and add them to the context, I want to find those classes in an external library, so I can't add annotations to them.
+
+Let's see how we can solve this issue by developing a custom extension.
+
+== Creating the extension
+
+Creating the extension is as simple as:
+
+[source, bash]
+----
+mvn io.quarkus:quarkus-maven-plugin:create-extension -DwithoutTests
+----
+
+It will ask for a `groupId` - let's keep the default `org.acme` - and an extension id - I went for `message-transformers-as-beans`.
+
+Then you can import your new extension into your favorite IDE.
+
+== Structure of the extension
+
+There is a lot to say about extensions but, in the context of this blog post, we will keep it short.
+The extension is composed of three Maven modules:
+
+* The parent module - nothing to see here
+* The deployment module - this is the one of interest for our blog post
+* The runtime module - in this blog post, we won't modify it
+
+Let's keep it simple: the deployment module is what will be used at build time, the runtime module is what will be used at runtime.
+
+In our case, we want to declare new beans and this is something we do at build time, so deployment module, here we come!
+
+== Processors and build steps
+
+If you have a look at your `deployment` module, you will see a `MessageTransformersAsBeansProcessor` and you can see a method annotated with the `@BuildStep` annotation in it.
+
+Quarkus build is populated by these build steps and they are following a consumer/producer model with dependency injection.
+The items being consumed and produced are called ``BuildItem``s.
```

Comment:
```suggestion
The items being consumed and produced are called `BuildItem`s.
```

Reply:
That's how you're supposed to do things when you have a character just after the backtick.

Reply:
But what add the backtick in the first place?